libcoap/src/oscore.rs

```diff
@@ -0,0 +1,101 @@
+        // SAFETY: Provided pointer and length point to a valid byte string usable by
+        // coap_new_str_const().
+        let conf = unsafe { coap_new_str_const(oscore_conf_bytes.as_ptr(), oscore_conf_bytes.len()) };
+        if conf.is_null() {
+            return Err(OscoreConfigError::Unknown);
+        }
+
+        // SAFETY:
+        // The parts of the byte string referenced by conf are defensively copied if used
+        // by the newly created oscore_conf.
+        // Conf was just checked for invalidity, whether or not it containes all required fields.
+        // - save_seq_num_func is specifically designed to work as a callback for this
+        //   function.
+        // - save_seq_num_func_param may be a null pointer (save_seq_num_func does
+        //   not use it).
+        let oscore_conf = unsafe { coap_new_oscore_conf(*conf, Some(save_seq_num_func), ptr::null_mut(), seq_initial) };
+        if oscore_conf.is_null() {
+            return Err(OscoreConfigError::Unknown);
+        }
+
+        // Save the initial recipient_id (if present). This needs to be added to the context when
+        // calling oscore_server to prevent a double free when trying to add an identical
+        // recipient_id later.
+        let mut initial_recipient: Option<String> = None;
+        let oscore_conf_str = core::str::from_utf8(oscore_conf_bytes).expect("could not parse config bytes to str");
+        for line in oscore_conf_str.lines() {
+            if line.starts_with("recipient_id") {
+                let parts: Vec<&str> = line.split(",").collect();
+                initial_recipient = Some(parts[2].trim().trim_matches('"').to_string());
+                break;
+            }
+        }
+
+        // Return the valid OscoreConf.
+        Ok(Self {
+            raw_conf: oscore_conf,
+            initial_recipient,
+            drop: true,
+        })
+    }
+
+    /// Cosumes the OscoreConf and returns the contained raw_conf libcoap struct as well as an
+    /// optional initial recipient if set.
+    /// SAFETY: Caller must assure to free of the raw_conf returned by this function using
+    /// coap_delete_oscore_conf() to prevent leaking memory.
+    pub(crate) unsafe fn into_raw_conf(mut self) -> (*mut coap_oscore_conf_t, Option<String>) {
+        self.drop = false;
+        (self.raw_conf, self.initial_recipient.clone())
+    }
+}
+
+impl Drop for OscoreConf {
+    /// Drop the OscoreConf's raw_conf.
+    fn drop(&mut self) {
+        if self.drop {
+            // SAFETY: If the CoapConf was consumed by calling the unsafe function into_raw_conf() the
+            // instance will be marked to not try to drop the raw_conf here as this should be handled
+            // by the caller. Only try to free the raw_conf if the instance was not consumed and
+            // therby not freed off before to prevent laeking memory.
+            unsafe {
+                coap_delete_oscore_conf(self.raw_conf);
+            }
+        }
+    }
+}
```

Comment:
We can reduce the size of the `OscoreConf` data structure and get rid of the `drop` attribute by replacing the pointer with a null pointer when destructuring into a raw object and checking for a null pointer on destruction.
(Note that you also have to remove the `drop` field from the constructor and type definition).

```suggestion
        // Replace pointer in structure with a null pointer, so the destructor knows that we know longer own
        // the raw structure and must therefore not free it.
        let raw_conf = std::mem::take(&mut self.raw_conf);
        (raw_conf, self.initial_recipient.clone())
    }
}

impl Drop for OscoreConf {
    /// Drop the OscoreConf's raw_conf.
    fn drop(&mut self) {
        if !self.raw_conf.is_null() {
            // SAFETY: If the CoapConf was consumed by calling the unsafe function into_raw_conf() the
            // pointer will be null, but we just checked that this is not the case.
            // Therefore, we are still the owner of the raw config and can therefore free it.
            unsafe {
                coap_delete_oscore_conf(self.raw_conf);
            }
        }
    }
}
```